apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient5-plugin/src/main/java/co/elastic/apm/agent/httpclient/v5/helper/AsyncRequestProducerWrapper.java

```diff
@@ -78,44 +81,42 @@ public void sendRequest(RequestChannel requestChannel, HttpContext httpContext)
 
     @Override
     public boolean isRepeatable() {
-        return delegate != null && delegate.isRepeatable();
+        return delegate.isRepeatable();
     }
 
     @Override
     public void failed(Exception e) {
-        if (delegate != null) {
+        try {
             delegate.failed(e);
+        } finally {
+            internalResetState();
         }
     }
 
     @Override
     public int available() {
-        return delegate != null ? delegate.available() : 0;
+        return delegate.available();
     }
 
     @Override
     public void produce(DataStreamChannel dataStreamChannel) throws IOException {
-        if (delegate != null) {
-            delegate.produce(dataStreamChannel);
-        }
+        // this method might be called after releaseResources, hence preventing us from implementing recycling easily
```

Comment:
[for reviewer] an early recycling did made `delegate` field `null`, which made the caller wait forever.